Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -86,39 +86,49 @@ public virtual void ProcessBlock(Block block)
 
             // If the new block previous hash is the same as the 
             // wallet hash then just pass the block to the manager. 
+            var newTip = this.chain.GetBlock(block.GetHash());
+
+            // new block is not in the main chain nothing more to do right now. 
+            if (newTip == null)
+                return; // reorg
+
             if (block.Header.HashPrevBlock != this.walletTip.HashBlock)
             {
-                // set a temporary wallet tip in case we are not on the best chain anymore.
-                var walletTip = this.WalletTip;
-
                 // If previous block does not match there might have 
                 // been a reorg, check if the wallet is still on the main chain.
                 ChainedBlock inBestChain = this.chain.GetBlock(this.walletTip.HashBlock);
                 if (inBestChain == null)
                 {
                     // The current wallet hash was not found on the main chain.
-                    // A reorg happenend so bring the wallet back top the last known fork.
+                    // A reorg happened so bring the wallet back top the last known fork.
 
-                    var fork = walletTip;
+                    var fork = this.walletTip;
 
                     // We walk back the chained block object to find the fork.
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
 
                     this.walletManager.RemoveBlocks(fork);
```

Comment:
What if anything goes wrong in the removeblocks method? We could have removed some uxto's in the walletmanager and failed halfway due to a concurrency issue. I still have to think of a scenario here but let's say it happened would a next block be able to repair it?

Reply:
Also would it make sense to create some tests for the mentioned scenario's discussed before my comment?

Reply:
If removeblocks fails that should throw an exception, remove blocks has a lock internally so it will not have race issues, the race condition is in regard to the chain being reorged (this method should (and will be synchronized once we push the fix to make consensus synchronised). 

Yes definitely need to write tests to cover each scenario.
Can you specify which is the failing test? 
I will follow your tests to create a test for each path.

Reply:
The failing test is mentioned in #570.